### ISSUE TITLE
Verify subnet definition in terraform

### ISSUE THREAD

**Issue by raquelalegre** [UCL-CloudLabs/user-portal#62]
We used to define the vnet and the subnet separately in our terraform template, like this:

```

# create virtual network
resource "azurerm_virtual_network" "vnet" {
    name = "tfvnet"
    address_space = ["10.0.0.0/16"]
    location = "ukwest"
    resource_group_name = "${azurerm_resource_group.rg.name}"
}

# create subnet
resource "azurerm_subnet" "subnet" {
    name = "tfsub"
    resource_group_name = "${azurerm_resource_group.rg.name}"
    virtual_network_name = "${azurerm_virtual_network.vnet.name}"
    address_prefix = "10.0.2.0/24"
    #network_security_group_id = "${azurerm_network_security_group.nsg.id}"
}

```

Since Azure changed the way virtual networks need to be defined in the last week, we now need to adapt our terraform template to include the subnet in the virtual network definition:

```

# create virtual network
resource "azurerm_virtual_network" "vnet" {
    name = "tfvnet"
    address_space = ["10.0.0.0/16"]
    location = "ukwest"
    resource_group_name = "${azurerm_resource_group.rg.name}"
    subnet {
        name = "tfsub"
        address_prefix = "10.0.2.0/24"

    }
}

# create subnet
resource "azurerm_subnet" "subnet" {
    name = "tfsub"
    resource_group_name = "${azurerm_resource_group.rg.name}"
    virtual_network_name = "${azurerm_virtual_network.vnet.name}"
    address_prefix = "10.0.2.0/24"
    #network_security_group_id = "${azurerm_network_security_group.nsg.id}"
}

```

The separate subnet defined  in the first example seems to be now redundant, but removing it from the terraform file will cause an error, since the network interface refers to the subnet id, and it won't exist anymore. Changing the reference from the network interface to `azure_virtual_network.vnet.subnet.id` doesn't work either, because Terraform will only create the `id` field behind the scenes for elements in the parent level. 

So it seems odd that we need to define the same subnet in two places. It also apparently work if the subnet name and address prefix are different in each element. Is this creating two separate subnets? Why is there not an error if they have the same name and address prefix then? This might be a glitch caused by terraform not keeping up quick enough with Azure changes in their API. 

We'll need to come back to this again in the future and check this doesn't cause any problems, and update the template if needed.

**Comment by ageorgou** [comment]
Just noting that the Terraform docs (currently, at least) [seem to suggest](https://www.terraform.io/docs/providers/azurerm/r/subnet.html) that we should not be doing this:

> NOTE on Virtual Networks and Subnet's: Terraform currently provides both a standalone Subnet resource, and allows for Subnets to be defined in-line within the Virtual Network resource. At this time you cannot use a Virtual Network with in-line Subnets in conjunction with any Subnet resources. Doing so will cause a conflict of Subnet configurations and will overwrite Subnet's.

So it seems like we are indeed creating a clash, but one of the definitions is just overwritten silently, which is why we're not getting any errors. Still not clear whether requiring the subnet as a "nested" resource is an intentional change.

**Comment by ageorgou** [comment]
It looks like we can work around this by using Terraform's [data sources](https://www.terraform.io/docs/configuration/data-sources.html). Replacing the explicit subnet resource with [something like this](https://www.terraform.io/docs/providers/azurerm/d/subnet.html) should work:

```
data "azurerm_subnet" "subnet" {
  name                 = "tfsub"  # or make this a variable to match the virtual net def
  virtual_network_name = "${azurerm_virtual_network.vnet.name}"
  resource_group_name  = "${azurerm_resource_group.rg.name}"
}
```

The network interface definition can then refer to `"${data.azurerm_subnet.subnet.id}"`.
We could also avoid hardcoding the subnet name by using a [virtual network data source](https://www.terraform.io/docs/providers/azurerm/d/virtual_network.html) and getting the subnet name that way.

I haven't tested this, but the docs on when data sources are created and updated make it sound like it should work (i.e. there are no circular dependencies).

**Comment by ageorgou** [comment]
Closed in #77.
